healthcheck/tor_connection.go

```diff
@@ -0,0 +1,85 @@
+package healthcheck
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"syscall"
+
+	"github.com/lightningnetwork/lnd/tor"
+)
+
+// CheckTorServiceStatus checks the Onion Service's info to decide whether the
+// Tor daemon is up and running. We will get an EOF or a broken pipe error if
+// the Tor daemon is stopped/restarted as the previously created socket
+// connection is no longer open. In this case, we will attempt a restart on our
+// tor controller. If the tor daemon comes back, a new socket connection will
+// then be created.
+func CheckTorServiceStatus(tc *tor.Controller, createService func() error) error {
+	err := tc.GetServiceInfo()
+	if err == nil {
+		return nil
+	}
+
+	log.Debugf("Checking tor service got: %v", err)
+
+	switch {
+	// We will get an EOF if the connection is lost. In this case, we will
+	// return an error and wait for the Tor daemon to come back. We won't
+	// attempt to make a new connection since we know Tor daemon is down.
+	case errors.Is(err, io.EOF), errors.Is(err, syscall.ECONNREFUSED):
+		log.Warnf("Tor connection lost, attempting a tor controller " +
```

Comment:
This log is a bit unintuitive, we might not actually attempt re-connection in the next check if tor continues to give us an EOF? Also thing that the healthcheck will log the error we return so we don't need to log and return here?

Reply:
I think we would only ever get one EOF, which signals the connection is lost. And yeah, if it keeps giving EOF, there's no point to re-connect because there's no socket available.

The warning log is removed now.

Reply:
A second thought. I think the healthcheck should give warning rather than info so the user can act upon it?